src/rm_cms.c

```diff
@@ -0,0 +1,309 @@
+            return RedisModule_ReplyWithError(ctx, "CMS: invalid value");  
+
+        *width = ceil(2 / err);
+        *depth = ceil(log10f(prob) / log10f(0.5));
+    }
+    return REDISMODULE_OK;
+}
+
+int CMSketch_create(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    if(argc != 4)
+        return RedisModule_WrongArity(ctx);
+
+    CMSketch *cms;
+    long long width, depth;
+    RedisModuleString *keyName = argv[1];
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ|REDISMODULE_WRITE);
+
+    if(parseCreateArgs(ctx, argv, argc, &width, &depth) != REDISMODULE_OK)
+        return REDISMODULE_ERR;
+
+    if(RedisModule_KeyType(key) != REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_CloseKey(key);
+        return RedisModule_ReplyWithError(ctx,"CMS: key already exists");
+    }
+
+    CreateCmsKey(ctx, keyName, width, depth, &cms, &key);
+    RedisModule_CloseKey(key);
+
+    RedisModule_Log(ctx, "info", "created new count-min sketch");
+    RedisModule_ReplyWithSimpleString(ctx, "OK");
+    RedisModule_ReplicateVerbatim(ctx);
+    return REDISMODULE_OK;
+}
+
+static int parseIncrByArgs(RedisModuleCtx *ctx, RedisModuleString **argv, int argc,
+                CMSPair **pairs, int qty) {
+    *pairs = (CMSPair *)CMS_CALLOC(qty, sizeof(CMSPair));
+    for(int i = 0; i < qty; ++i) {
+        pairs[i]->key = RedisModule_StringPtrLen(argv[2 + i * 2], NULL);
+        RedisModule_StringToLongLong(argv[2 + i * 2 + 1], &(pairs[i]->value));
+    }
+    return REDISMODULE_OK;
+}
+
+int CMSketch_incrBy(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx);
+
+   if(argc < 4 || (argc % 2) == 1) {
+        return RedisModule_WrongArity(ctx);
+    }
+
+    int pairCount = (argc - 2) / 2;
+    RedisModuleString *keyName = argv[1];
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ);
+    CMSketch *cms;
+    CMSPair *pairArray;
+    parseIncrByArgs(ctx, argv, argc, &pairArray, pairCount);
+
+    if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        CreateCmsKey(ctx, keyName, DEFAULT_WIDTH, DEFAULT_DEPTH, &cms, &key);
+    } else if(RedisModule_ModuleTypeGetType(key) != CMSketchType) {
+        return RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
+    } else {
+        cms = RedisModule_ModuleTypeGetValue(key);
+    }
+ 
+    for(int i = 0; i < pairCount; ++i) {
+        CMS_IncrBy(cms,pairArray[i].key, pairArray[i].value);
+    }    
+
+    CMS_FREE(pairArray);
+    RedisModule_ReplyWithSimpleString(ctx, "OK");
+    RedisModule_CloseKey(key);
+    return REDISMODULE_OK;
+}
+
+int CMSketch_query(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+    RedisModule_AutoMemory(ctx);
+
+    if(argc < 3) {
+        return RedisModule_WrongArity(ctx);
+    }
+
+    CMSketch *cms = NULL;
+    if(GetCMS(ctx, argv[1], &cms, REDISMODULE_READ) != REDISMODULE_OK) {
+        return REDISMODULE_ERR;
```

Comment:
Missing RM_CloseKey, no need to return REDISMODULE_ERR

Reply:
The closekey isn't needed as I called RedisModule_AutoMemory

Reply:
👍